NString.Tests/StringTemplateConverterTests.cs

```diff
@@ -0,0 +1,335 @@
+        [InlineData("hello", false)]
+        public void Generic_Converter_Allows_Null_Values(object input, bool expectedResult)
+        {
+            var values = new MiscConvertersValues { NullToBool = input };
+
+            string actual = StringTemplate.Format("{NullToBool}", values);
+
+            Assert.Equal(expectedResult, bool.Parse(actual));
+        }
+
+        [Fact]
+        public void Generic_Converter_CanConvert_Allows_Derived_Type()
+        {
+            var values = new MiscConvertersValues { String_Array_With_ReadOnlyCollection_Converter = StringCollectionValues };
+
+            string actual = StringTemplate.Format("{String_Array_With_ReadOnlyCollection_Converter}", values);
+
+            Assert.Equal(ExpectedStringCollectionConcatenation, actual);
+        }
+
+        [Fact]
+        public void Generic_Converter_Throws_When_Given_Incompatible_Value()
+        {
+            IStringTemplateValueConverter converter = new StringCollectionConverter();
+
+            var shouldWork = converter.Convert(StringCollectionValues);
+
+            Assert.Equal(ExpectedStringCollectionConcatenation, shouldWork); // sanity check
+
+            Assert.Throws<NotSupportedException>(() => converter.Convert("whatever"));
+        }
+
+        class StringCollectionConverter : StringTemplateValueConverter<IReadOnlyCollection<string>>
+        {
+            public override object Convert(IReadOnlyCollection<string> value) => string.Join("|", value);
+        }
+
+        class DateTimeToTicksConverter : StringTemplateValueConverter<DateTime>
+        {
+            public override object Convert(DateTime value) => value.Ticks;
+        }
+
+        class AddFortyTwoDaysConverter : StringTemplateValueConverter<DateTime>
+        {
+            public override object Convert(DateTime value) => value + TimeSpan.FromDays(42);
+        }
+
+        class GetDayNameConverter : StringTemplateValueConverter<DateTime>
+        {
+            public override object Convert(DateTime value) => value.DayOfWeek.ToString();
+        }
+
+        class NullToBoolConverter : StringTemplateValueConverter<object>
+        {
+            public override object Convert(object value) => value == null;
+        }
+
+        class IncrementStateConverter : IStringTemplateValueConverter
+        {
+            public int State;
+            public bool CanConvert(Type objectType) => true;
+
+            public object Convert(object value) => Interlocked.Increment(ref State);
+        }
+
+        class CanNeverConvertConverter : IStringTemplateValueConverter
+        {
+            public bool CanConvert(Type objectType) => false;
+
+            public object Convert(object value) => throw new NotSupportedException();
+        }
+
+        class PropertyOrFieldValues
+        {
+            [StringTemplateValueConverter(typeof(StringCollectionConverter))]
+            public IReadOnlyCollection<string> ConcatenatedStringCollectionProperty { get; set; }
```

Comment:
Nitpick: could you please name properties based on what they are, rather than how they're converted?

Reply:
I'll try, but some properties only differ between each other based on their converters, so it might get a bit ugly.

Reply:
I'm not sure it's better, tell me what you think.
(maybe it would be better with underscore separated words even in properties names?)

Reply:
Not ideal, but I don't really have a better idea... Let's keep it like this.